codalab/rest/bundles.py

```diff
@@ -451,16 +458,79 @@ def _fetch_bundle_locations(bundle_uuid: str):
 )
 def _add_bundle_location(bundle_uuid: str):
     """
-    Adds a new BundleLocation to a bundle. Request body must contain the fields in BundleLocationSchema.
+    Adds a new BundleLocation to a bundle. If need to generate sas token, generate Azure SAS token and connection string. Request body must contain the fields in BundleLocationSchema.
 
     Query parameters:
-    - `bundle_uuid`: Bundle UUID corresponding to the new location
+    - `need_sas`: (Optional) Bool. If true, if will return SAS token
+    - `is_dir`: (Optional) Bool. Whether the uploaded file is directory.
     """
+    check_bundles_have_all_permission(local.model, request.user, [bundle_uuid])
+    need_sas = query_get_bool('need_sas', default=False)
+    is_dir = query_get_bool('is_dir', default=False)
+
+    bundle = local.model.get_bundle(bundle_uuid)
     new_location = BundleLocationSchema(many=True).load(request.json).data[0]
-    new_location["uuid"] = local.model.add_bundle_location(
-        new_location['bundle_uuid'], new_location['bundle_store_uuid']
+    logging.info(
+        f"Try to bypass server upload, need_sas: {need_sas}, is_dir: {is_dir}, new_location: {new_location}"
     )
-    return BundleLocationSchema(many=True).dump([new_location]).data
+    # Scenario 1: User does not sepcify destination store, but rest-server set default storage name.
+    # Should bypass server and upload to default Azure store.
+    if (
+        new_location.get('bundle_store_uuid', None) is None
+        and os.environ.get('CODALAB_DEFAULT_BUNDLE_STORE_NAME') is not None
```

Comment:
We can replace this line and the next line with just:

```suggestion
        and os.environ.get('CODALAB_DEFAULT_BUNDLE_STORE_NAME')
```